### PR TITLE
DOC: Generalize "rename" documentation

### PR DESCRIPTION
**Pull request by MaxVanDeursen**
In this pull request, the documentation of the rename function of NDFrame is generalized over all its children using the function: DataFrame, Series and Panel. These three classes now use the Substitute decorator to add the generic part of the documentation to their own, using the same decorator to replace class specific parameters into the generic documentation.

I have tried my best to create as little changes to the documentation of each individual function. Although I have tried to get rid of all style errors, I do still receive some errors from the `validate_docstrings.py`:

```
Series
2 Errors found:
	Parameters {**kwargs} not documented
	Unknown parameters {errors, copy, inplace, level}
```

```
Panel
2 Errors found:
	Parameters {minor_axis, items, **kwargs, major_axis} not documented
	Unknown parameters {items, major_axis, minor_axis, copy, errors, inplace, level}
```

Help is appreciated!

### CODE REVIEW

**Review by WillAyd — CHANGES_REQUESTED** [review]
Thanks for the PR! Per our conversation on Gitter I think you can actually move a lot more into the generic docstring and reduce the amount of substitution going on here. I think this is confused by the current Series docstring which may not be properly documenting all of the allowed parameters.

If you don't mind looking a little further and seeing if we can simplify by moving more into the generic docstring would be greatly appreciated

**Comment by MaxVanDeursen** [comment]
From our conversation yesterday and from the code, you would think that indeed the `mapper`, `columns`, and `axis` parameters are allowed as well.

Upon further inspection, I did notice that they are indeed not disallowed. However, they do not seem to have the same behavior as the DataFrame counterparts:
```python
s = pd.Series([1,2,3,4])
df = pd.DataFrame([1,2,3,4])

s
0    1
1    2
2    3
3    4
dtype: int64

df
   0
0  1
1  2
2  3
3  4

s.rename(mapper={0:'x'}, axis=0)
0    1
1    2
2    3
3    4
dtype: int64

df.rename(mapper={0:'x'}, axis=0)
   0
x  1
1  2
2  3
3  4

s.rename(columns={"NONEXISTENT": 'a'}, errors="raise")
0    1
1    2
2    3
3    4
dtype: int64

df.rename(colums={"NONEXISTENT": 'a'}, errors="raise")
Traceback (most recent call last):
...
KeyError: "['NONEXISTENT'] not found in axis"

```

From this, I do not think it is a good decision to generalize the documentation over these arguments, as they do not provide the same functionality for Series and DataFrame. Generalizing the Series documentation to essentially the one of DataFrame, leads to a misleading documentation for Series.

**Comment by MaxVanDeursen** [comment]
@WillAyd Ping

**Comment by WillAyd** [comment]
Can you take a look at reindex? I think that has a similar issue from an API perspective but handles the docstring substitution in a way that requires less duplication

**Comment by MaxVanDeursen** [comment]
I have now tried to use the `_shared_doc_kwargs` together with a few additions:

- `altered`: To specify what is being altered
- `axes_alt_types`: To specify the alternative types of the axes, which are different for Series since it also includes the rename use of `Series.name`
- `alternative_use`: A string describing the above alternative use.

I have not found a way to effectively generalize the `mapper` parameter for DataFrame, as there is no such parameter for Series. I have thought of three options so far:
1. Include another addition to the substitute, which describes the following: `Instead, ``mapper`` may be used together with ``axis`` to apply the function to the axis values.'`. Notice that this would be only for the DataFrame, and left empty for Series (and Panel). This is similar to `alternative_use` and `axes_alt_types`
2. Include the same description of 1, only in `alternative_use`. However, `alternative_use` is used in the function description as well, hence making a bit of an illogical description. A solution to this would be to remove `alternative_use` from this description, however, this would make the description of Series only include the `Series.name` explanation in it's `index` parameter description.

**Comment by MaxVanDeursen** [comment]
@WillAyd Any ideas?

**Comment by TomAugspurger** [comment]
It sounds like there are potentially two issues here:

1. Reduce the duplication of docs between Series.rename and DataFrame.rename (lets ignore panel, it's going away)
2. Differences in behavior between Series.rename and DataFrame.rename

I would isolate number 2 first. For example


```
s.rename(columns={"NONEXISTENT": 'a'}, errors="raise")
```

Seems buggy. I think providing non-None `columns` to Series.rename should raise.

I think once we agree on the correct behavior it'll be easier to document and determine whether or not the  docstrings should be shared.

**Comment by MaxVanDeursen** [comment]
I see what you mean. I do not think I am the one to decide on the behavior of `Series.rename`!

> Seems buggy. I think providing non-N�one columns to Series.rename should raise.
Even None columns would not make sense though, in the context of Series, since Series does not have columns.

To be honest, I would say to either force the API of Series.rename (i.e. throw on any non-documented argument) or make sure all possible arguments of NDFrame.rename make sense to Series as well, and provide behaviour.

**Review by WillAyd — CHANGES_REQUESTED** [review]

File: pandas/core/frame.py [hunk 1]
```diff
@@ -3913,62 +3913,13 @@ def drop(self, labels=None, axis=0, index=None, columns=None,
                                              ('inplace', False),
                                              ('level', None),
                                              ('errors', 'ignore')])
+    @Substitution(**dict(_shared_doc_kwargs,
```
[the remark this answers is not in the document]
Comment (MaxVanDeursen) [reply pandas/core/frame.py]: I do think that that would be better; however, when splitting this into multiple Substitution decorations, the "top" Substitution is not substituted into the `NDFrame.rename` documentation.
For example; if we use the following Substitution decorations:
```python
    @Substitution(**_shared_doc_kwargs)
    @Substitution(altered='axes input function or functions',
                            alternative_use='',
                            axes_alt_types='')
    @Substitution(generic_documentation=NDFrame.rename.__doc__)
```

I get the following error when building the documentation:
```Traceback (most recent call last):
  File "make.py", line 341, in <module>
    sys.exit(main())
  File "make.py", line 329, in main
    globals()['pandas'] = importlib.import_module('pandas')
  File "/opt/anaconda3/lib/python3.7/importlib/__init__.py", line 127, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "<frozen importlib._bootstrap>", line 1006, in _gcd_import
  File "<frozen importlib._bootstrap>", line 983, in _find_and_load
  File "<frozen importlib._bootstrap>", line 967, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 677, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 728, in exec_module
  File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
  File "/home/max/git/pandas/pandas/__init__.py", line 42, in <module>
    from pandas.core.api import *
  File "/home/max/git/pandas/pandas/core/api.py", line 40, in <module>
    from pandas.core.panel import Panel
  File "/home/max/git/pandas/pandas/core/panel.py", line 109, in <module>
    class Panel(NDFrame):
  File "/home/max/git/pandas/pandas/core/panel.py", line 1257, in Panel
    def rename(self, items=None, major_axis=None, minor_axis=None, **kwargs):
  File "/home/max/git/pandas/pandas/util/_decorators.py", line 259, in __call__
    func.__doc__ = func.__doc__ and func.__doc__ % self.params
KeyError: 'axes'
```

This is what I encountered before, and I found out the pushed code fixes this issue.

File: pandas/core/panel.py [hunk 2]
```diff
@@ -1249,9 +1249,83 @@ def reindex(self, *args, **kwargs):
             result = super(Panel, self).reindex(**kwargs)
         return result
 
-    @Substitution(**_shared_doc_kwargs)
-    @Appender(NDFrame.rename.__doc__)
+    @Substitution(**dict(_shared_doc_kwargs,
```
[the remark this answers is not in the document]
Comment (MaxVanDeursen) [reply pandas/core/panel.py]: I understand; however, it still had to be made compliant with the `NDFrame.generic` substitutions in order for the documentation to be built (without errors).

**Comment by jreback** [comment]
cc @jorisvandenbossche @datapythonista

**Comment by jreback** [comment]
good idea, but closing as stale

**Pull request closed without merging**